lib/auth/oidc/provider.go

```diff
@@ -0,0 +1,189 @@
+// will start up background resources to manage the cache.
+func NewProviderCache() *ProviderCache {
+
+	// Create a context, so a server that is shutting down can correctly
+	// shut down the cache loop and OIDC provider background processes.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &ProviderCache{
+		providers: map[string]*oidc.Provider{},
+		cancel:    cancel,
+	}
+
+	// Start the cleanup timer
+	go result.runCleanupLoop(ctx)
+
+	return result
+}
+
+// Get returns the OIDC provider for the given auth method configuration.
+// This will initialize the provider if it isn't already in the cache or
+// if the configuration changed.
+func (c *ProviderCache) Get(authMethod *structs.ACLAuthMethod) (*oidc.Provider, error) {
+
+	// No matter what we'll use the config of the arg method since we'll
+	// use it to compare to existing (if exists) or initialize a new provider.
+	oidcCfg, err := providerConfig(authMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get any current provider for the named auth-method.
+	var (
+		current *oidc.Provider
+		ok      bool
+	)
+
+	// This may be the first time the cache is being used, therefore we need to
+	// ensure we don't panic when attempting to access the map.
+	c.mu.RLock()
+	if c.providers != nil {
+		current, ok = c.providers[authMethod.Name]
+	}
+	c.mu.RUnlock()
+
+	// If we have a current value, we want to compare hashes to detect changes.
+	if ok {
+		currentHash, err := current.ConfigHash()
+		if err != nil {
+			return nil, err
+		}
+
+		newHash, err := oidcCfg.Hash()
+		if err != nil {
+			return nil, err
+		}
+
+		// If the hashes match, this is can be classed as a cache hit.
+		if currentHash == newHash {
+			return current, nil
+		}
+	}
+
+	// If we made it here, the provider isn't in the cache OR the config
+	// changed. We therefore, need to initialize a new provider.
+	newProvider, err := oidc.NewProvider(oidcCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// If we have an old provider, clean up resources.
+	if current != nil {
+		current.Done()
+	}
+
+	c.providers[authMethod.Name] = newProvider
+
+	return newProvider, nil
+}
+
+// Delete force deletes a single auth method from the cache by name.
+func (c *ProviderCache) Delete(name string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	p, ok := c.providers[name]
+	if ok {
+		p.Done()
+		delete(c.providers, name)
+	}
+}
+
+// Shutdown stops any long-lived cache process and informs each OIDC provider
+// that they are done. This should be called whenever the Nomad server is
+// shutting down.
+func (c *ProviderCache) Shutdown() {
+	c.cancel()
+	c.clear()
+}
+
+// runCleanupLoop runs an infinite loop that clears the cache every cacheExpiry
+// duration. This ensures that we force refresh our provider info periodically
+// in case anything changes.
+func (c *ProviderCache) runCleanupLoop(ctx context.Context) {
```

Comment:
We don't need to do it in this PR but we might want to consider whether the RPC handlers can/should force a cache flush of the provider if they hit errors that are caused by the upstream provider (ex. this is a very long lived HTTP client and I don't know how `cap` has its TCP connections configured to handle a DNS change at the provider; ref https://github.com/golang/go/issues/23427)

Reply:
Great shout; I'll note this for future thought and raise as a followup issue.